lsif_indexer/indexer.py

```diff
@@ -0,0 +1,439 @@
+        to_index = set(self.documents.values())
+        with Timer("Indexing", 0) as timer:
+            yield from self.toposort_index(to_index)
+            for document in self.documents.values():
+                yield from document.emit_references()
+            timer.description = f"Indexed {len(self.documents)} files"
+        document_ids = list(
+            sorted({document.id for document in self.documents.values()})
+        )
+        yield model.Contains(out_v=self.project_id, in_vs=document_ids)
+        yield model.End(scope="project", data=self.project_id)
+
+    def toposort_index(self, to_index: set[Document]):
+        stack: list[Path] = []
+
+        def try_to_index(document: Document):
+            stack.append(document.filename)
+            for dep in document.get_dependencies(self):
+                if not dep.indexed:
+                    if dep.filename in stack:
+                        if (
+                            dep.filename.name == "__init__.py"
+                            and dep.filename.parent
+                            in document.filename.parents
+                        ):
+                            # Always let a parent or sibling __init__.py
+                            # go last, people often have circular
+                            # dependencies in packages and their
+                            # modules, we should ignore that and just
+                            # index what we can.
+                            continue
+                        raise CircularDependencyError(stack, dep.filename)
+                    to_index.discard(dep)
+                    yield from try_to_index(dep)
+                assert dep.indexed
+            yield from document.index(self)
+            stack.pop()
+
+        while to_index:
+            document = to_index.pop()
+            yield from try_to_index(document)
+
+    def load_files(self):
+        filenames = self.root.glob("**/*.py")
+        filenames = (
+            filename
+            for filename in filenames
+            if not any(d in filename.parents for d in self.exclude_dir)
+        )
+        scripts = [
+            jedi.Script(path=filename, project=self.project)
+            for filename in filenames
+        ]
+        for script in scripts:
+            document = Document(script=script)
+            self.documents[document.filename] = document
+
+    def get_document(self, filename: Path) -> Optional[Document]:
+        return self.documents.get(filename)
+
+
+@click.command()
+@click.argument(
+    "root",
+    type=click.Path(
+        exists=True,
+        file_okay=False,
+        dir_okay=True,
+        readable=True,
+        resolve_path=True,
+        allow_dash=False,
+    ),
+)
+@click.option("--file", type=click.File(mode="w"), default="data.lsif")
+@click.option("--verbose/--quiet")
+@click.option(
+    "--exclude-dir", type=click.Path(resolve_path=True), multiple=True
```

Comment:
TODO: change this to exclude things excluded from version control by default (like `venv`).